plugins/inputs/openstack/openstack.go

```diff
@@ -0,0 +1,1059 @@
+		return fmt.Errorf("unable to create V3 identity client: %v", err)
+	}
+
+	o.services = serviceMap{}
+	o.gatherServices()
+
+	if o.compute, err = openstack.NewComputeV2(provider, gophercloud.EndpointOpts{}); err != nil {
+		return fmt.Errorf("unable to create V2 compute client: %v", err)
+	}
+
+	// Create required clients and attach to the OpenStack struct
+	if o.network, err = openstack.NewNetworkV2(provider, gophercloud.EndpointOpts{}); err != nil {
+		return fmt.Errorf("unable to create V2 network client: %v", err)
+	}
+
+	// The Orchestration service is optional
+	if o.services.ContainsService(Orchestration) {
+		if o.stack, err = openstack.NewOrchestrationV1(provider, gophercloud.EndpointOpts{}); err != nil {
+			return fmt.Errorf("unable to create V1 stack client: %v", err)
+		}
+	}
+
+	// The Cinder volume storage service is optional
+	if o.services.ContainsService(volumeV2Service) {
+		if o.volume, err = openstack.NewBlockStorageV2(provider, gophercloud.EndpointOpts{}); err != nil {
+			return fmt.Errorf("unable to create V2 volume client: %v", err)
+		}
+	}
+
+	// Initialize resource maps and slices
+	o.projects = projectMap{}
+	o.hypervisors = hypervisorMap{}
+	o.flavors = flavorMap{}
+	o.servers = serverMap{}
+	o.volumes = volumeMap{}
+	o.storagePools = storagePoolMap{}
+	o.subnets = subnetMap{}
+	o.networks = networkMap{}
+	o.ports = portMap{}
+	o.aggregates = aggregateMap{}
+	o.nova_services = nova_serviceMap{}
+	o.agents = agentMap{}
+	o.diag = serverDiag{}
+	return nil
+}
+
+// gatherStacks collects stacks from the OpenStack API.
+func (o *OpenStack) gatherStacks() error {
+	page, err := stacks.List(o.stack, &stacks.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list stacks: %v", err)
+	}
+	stacks, err := stacks.ExtractStacks(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract stacks")
+	}
+	for _, stack := range stacks {
+		o.stacks[stack.ID] = stack
+	}
+	return nil
+}
+
+// gatherServices collects services from the OpenStack API.
+func (o *OpenStack) gatherServices() error {
+	page, err := services.List(o.identity, &services.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list services: %v", err)
+	}
+	services, err := services.ExtractServices(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract services")
+	}
+	for _, service := range services {
+		o.services[service.ID] = service
+	}
+	return nil
+}
+
+// gathernova_services collects nova_services from the OpenStack API.
+func (o *OpenStack) gatherNovaServices() error {
+	page, err := nova_services.List(o.compute, &nova_services.ListOpts{}).AllPages()
+	if err != nil {
+		return fmt.Errorf("unable to list nova_services: %v", err)
+	}
+	nova_services, err := nova_services.ExtractServices(page)
+	if err != nil {
+		return fmt.Errorf("unable to extract nova_services")
+	}
+	for _, nova_service := range nova_services {
+		o.nova_services[nova_service.ID] = nova_service
+	}
+	return nil
+}
+
```

Comment:
All those function look very similar, you first call `List` with empty options, then `AllPages()`, `ExtractServices` etc. The only difference is the used OpenStack member variables. The fields `services`, `nova_services`, etc. are even of the same interface type.

So how about unifying these functions and implement them only once. Return a map and set the respective members outside of this function (e.g. in `Gather()`) if you need to do this at all...

Reply:
I now see that you cannot unify the gather functions as those are referencing the packages directly instead of "class" instances.